tally/src/scope_impl.cc

```diff
@@ -0,0 +1,251 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include "tally/src/scope_impl.h"
+
+#include <algorithm>
+#include <map>
+#include <mutex>
+#include <sstream>
+#include <string>
+#include <thread>
+#include <utility>
+#include <vector>
+
+#include "tally/scope_builder.h"
+#include "tally/src/capable_of.h"
+#include "tally/src/noop_stats_reporter.h"
+
+namespace tally {
+
+ScopeImpl::ScopeImpl(const std::string &prefix, const std::string &separator,
+                     const std::map<std::string, std::string> &tags,
+                     std::chrono::seconds interval,
+                     std::shared_ptr<StatsReporter> reporter)
+    : prefix_(prefix),
+      separator_(separator),
+      tags_(tags),
+      interval_(interval),
+      reporter_((reporter == nullptr) ? NoopStatsReporter::New() : reporter),
+      running_(false) {
+  if (interval > std::chrono::seconds(0)) {
+    thread_ = std::thread(&ScopeImpl::Run, this);
+    running_ = true;
```

Comment:
note: this is race-y - set `running_` first

Reply:
Good point, will fix!